components/style/cgt_arena.rs

```diff
@@ -0,0 +1,361 @@
+///! generation of the arena is incremented and subsequent lookups for previously-
+///! allocated tokens will fail.
+///!
+///! The arena is implemented with a two-tiered allocation hierarchy: `ChunkList`s
+///! contain pointers to lazily-allocated `Chunk`s, which are arrays of `Item`s
+///! (wrapped in `AtomicRefCell`s). `ChunkList`s also contain a pointer to the next
+///! `ChunkList`. The first `ChunkList` is allocated inline in the `Arena` struct, and
+///! subsequent `ChunkList`s are lazily allocated. A given `ChunkList` (indirectly)
+///! contains CHUNKS_PER_LIST * ITEMS_PER_CHUNK items, and we try to make this
+///! number large enough so that most workloads should need little to no pointer
+///! chasing.
+///!
+///! The lockless coherency scheme is based on the following two properties:
+///!   (1) Lazy allocation of buffers (`Chunk`s or `ChunkList`s) is the responsibility
+///!       of whichever thread allocates the first item (i.e. performs the fetch_add
+///!       that results in an index mapping to the first slot) in the new buffer.
+///!       Other threads which concurrently allocate subsequent items will spin until
+///!       the buffer appears.
+///!   (2) Once a buffer is allocated, it can only be deallocated when the Arena is
+///!       in exclusive mode, which means that no thread can simultaneously attempt
+///!       to allocate or dereference tokens.
+
+/// Core arena structure. Callers always access the arena using an `ArenaArc`,
+/// which provides |Sync| refcounted access with shared/exclusive borrow semantics.
+pub struct Arena<Item: Default> {
+    generation: u32,
+    current_index: AtomicIsize,
+    chunk_list: ChunkList<Item>,
+    self_weak: ArenaWeak<Item>,
+}
+
+pub type ArenaArc<Item> = Arc<AtomicRefCell<Arena<Item>>>;
+type ArenaWeak<Item> = Weak<AtomicRefCell<Arena<Item>>>;
+
+impl<Item: Default> Arena<Item> {
+    /// Creates a new arena.
+    pub fn new() -> ArenaArc<Item> {
+        let result = Arc::new(AtomicRefCell::new(Arena {
+            generation: 1,
+            current_index: AtomicIsize::new(0),
+            chunk_list: ChunkList::<Item>::new(),
+            self_weak: Weak::new(),
+        }));
+
+        // Add a weak edge from the Arena to the containing Arc to facilitate
+        // Token::bind().
+        result.borrow_mut().self_weak = Arc::downgrade(&result);
+
+        result
+    }
+
+    /// Allocates a `Token`.
+    pub fn allocate(&self) -> Token<Item> {
+        let data = self.allocate_raw();
+        let arena = self.self_weak.upgrade().unwrap();
+        unsafe { Token::create(data, arena).unwrap() }
+    }
+
+    /// Allocates a token directly as a `TokenData`, which saves a few borrow
+    /// operations for callers that don't need to dereference the result.
+    pub fn allocate_raw(&self) -> TokenData {
+        let idx_signed = self.current_index.fetch_add(1, Ordering::Relaxed);
+        if idx_signed < 0 {
```

Comment:
Wasn't this undefined behavior? Could we instead use `add_fetch` and check `current_index == 0`? Also, you're using `isize`, which means than on 64-bit you will overflow in the integer conversion below before this point.